### PR TITLE
Entities Card: Add Header & Footer Editor

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

7b005160e3f9732255c31b4e747283b3

## Type of change

- [x] New feature (thank you!)

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request: TODO

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by bramkragten** [comment]
Thinking out loud here:

We are going to get action editors, row editors (+ entity detail editor), and then we have a separate tab for the header and footer.

Should we do the header and footer the same way?
Do we have a mockup of how the complete picture will look?

**Comment by zsarnett** [comment]
The problem is putting too much on one Tab. Header and footer (while maybe Advanced is a bad tab name) aren't something that will be on every card. 

Maybe we rename the tab to `Extra` or something 🤷‍♂️ 

We should not muddy up the row UI with more and more options that arent used in every card.

Same with the sensor card. Extra tab with the primary color field

**Comment by zsarnett** [comment]
But also, no I dont have a mock up. My mockups are made by CSS and HTML. I am somewhat dumb when it comes to editing software. idk why. Honestly I don't have a good finalized idea of how I want the entity detail editor and special row "adder" to look like yet. I am building them in steps, hence the other special row pr

**Comment by zsarnett** [comment]
I think I will come back to this once I have finalized the other editor for entities. So that everything is the same way

**Comment by SeanPM5** [comment]
> Honestly I don't have a good finalized idea of how I want the entity detail editor and special row "adder" to look like yet.

For adding special rows, my idea was to have it inside the Entity dropdowns themselves. So at the very top of the dropdown would be something like this (special row name + descriptive text):

image

And then under those, in the same dropdown list, would be all your entities just like today. This would eliminate the need for another button or tab to add special rows. I had a bunch more mockups for this but can't find them anymore.

**Comment by zsarnett** [comment]
Gif in the OP has been updated to reflect changes

File: src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts [hunk 1]
```diff
@@ -174,93 +143,90 @@ export class HuiEntitiesCardEditor extends LitElement
             ></ha-switch>
           </ha-formfield>
         </div>
+        <div class="side-by-side">
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts:146]: I think these should not be side by side, but both on their own row

File: src/panels/lovelace/editor/header-footer-editor/hui-dialog-create-headerfooter.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+  }
+
+  private async _renderHeaderFooterElement(
+    headerFooter: HeaderFooter
+  ): Promise<TemplateResult> {
+    const { type, icon } = headerFooter;
+    let config: LovelaceHeaderFooterConfig = { type };
+
+    if (this.hass) {
+      config = await getHeaderFooterStubConfig(
```
Comment (bramkragten) [inline src/panels/lovelace/editor/header-footer-editor/hui-dialog-create-headerfooter.ts:96]: We don't use this in the overview, so we should only do this when the user picks a header/footer.

We should add entities in a future PR, so we can render them and add support for custom cards probably.

File: src/panels/lovelace/editor/hui-element-editor.ts [hunk 3]
```diff
@@ -58,25 +50,19 @@ export interface UIConfigChangedEvent extends Event {
...
   @property({ attribute: false }) public hass!: HomeAssistant;
 
   @property({ attribute: false }) public lovelace?: LovelaceConfig;
 
-  @property() public elementType: "row" | "card" = "card";
-
   @internalProperty() private _yaml?: string;
 
-  @internalProperty() private _config?: LovelaceCardConfig | LovelaceRowConfig;
+  @internalProperty() private _config?: any;
```
Comment (bramkragten) [inline src/panels/lovelace/editor/hui-element-editor.ts:61]:
```suggestion
  @internalProperty() private _config?: T;
```

File: src/panels/lovelace/editor/hui-element-editor.ts [hunk 4]
```diff
@@ -58,25 +50,19 @@ export interface UIConfigChangedEvent extends Event {
   };
 }
 
-const GENERIC_ROW_TYPE = "generic-row";
-
 @customElement("hui-element-editor")
 export class HuiElementEditor extends LitElement {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/hui-element-editor.ts:54]:
```suggestion
export class HuiElementEditor<T> extends LitElement {
```

File: src/panels/lovelace/editor/hui-element-editor.ts [hunk 5]
```diff
@@ -108,11 +94,11 @@ export class HuiElementEditor extends LitElement {
     this._setConfig();
   }
 
-  public get value(): LovelaceCardConfig | LovelaceRowConfig | undefined {
+  public get value(): any | undefined {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/hui-element-editor.ts:97]:
```suggestion
  public get value(): T | undefined {
```

File: src/panels/lovelace/editor/hui-element-editor.ts [hunk 6]
```diff
@@ -108,11 +94,11 @@ export class HuiElementEditor extends LitElement {
...
     return this._config;
   }
 
-  public set value(config: LovelaceCardConfig | LovelaceRowConfig | undefined) {
+  public set value(config: any | undefined) {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/hui-element-editor.ts:101]:
```suggestion
  public set value(config: T | undefined) {
```

File: src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts [hunk 7]
```diff
@@ -105,28 +87,15 @@ export class HuiEntitiesCardEditor extends LitElement
...
-    if (this._editRowConfig) {
+    if (this._subElementEditorConfig) {
       return html`
-        <hui-detail-editor-base
+        <hui-sub-element-editor
           .hass=${this.hass}
-          .guiModeAvailable=${this._editRowGuiModeAvailable}
-          .guiMode=${this._editRowGuiMode}
-          @toggle-gui-mode=${this._toggleMode}
+          .config=${this._subElementEditorConfig}
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts:94]: This doesn't work, as this object will not update when it is changed in the editor. So changing from the visual editor to the code editor will always use to the old config. (and the `hui-sub-element-editor` should not manage the config, it should be passed on update)

File: src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts [hunk 8]
```diff
@@ -105,28 +87,15 @@ export class HuiEntitiesCardEditor extends LitElement
...
           @go-back=${this._goBack}
+          @config-changed=${this._valueChanged}
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts:96]: Let's not use `_valueChanged` here, that function is getting really complicated and unreadable now. Create a new function for the sub config handling.

File: src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts (hunk 1 above)
Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts:146]: hmm they dont really take up that much room... but sure

File: src/panels/lovelace/editor/header-footer-editor/hui-dialog-create-headerfooter.ts (hunk 2 above)
Comment (zsarnett) [reply src/panels/lovelace/editor/header-footer-editor/hui-dialog-create-headerfooter.ts:96]: Maybe. I like the simplistic icons as well :)

File: src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts (hunk 7 above)
Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts:94]: Not sure I understand. The config is updated in the `_handleSubElementChanged`
```ts
this._subElementEditorConfig!.elementConfig = value;
```

Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts:94]: `@config-changed` event that happens from the `hui-row-element-editor` and the `hui-headerfooter-element-editor` actually passes through the `hui-sub-element-editor` and into the card editor.

Comment (zsarnett) [reply src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts:94]: `hui-sub-element-editor.ts`
```ts
private _handleConfigChanged(ev: CustomEvent): void {
  this._guiModeAvailable = ev.detail.guiModeAvailable;
}
```

Comment (bramkragten) [reply src/panels/lovelace/editor/config-elements/hui-entities-card-editor.ts:94]: Try editing something in the UI editor and then change it to the code editor, the code editor config will not be updated.

**Comment by zsarnett** [comment]
I don't think I understand the T typing...

**Comment by bramkragten** [comment]
> I don't think I understand the T typing...

OK, ask me on discord if you're ready

**Review by bramkragten — APPROVED** [review]

**Pull request merged**